File: nowcoder/二叉搜索树的第k个结点.py
# encoding: utf-8

'''

@author: liangchi

@contact: bnu_llc@163.com

@software: pycharm

@file: 二叉搜索树的第k个结点.py

@time: 2018/6/9 下午10:04

'''


class Solution:
    # 返回对应节点TreeNode

    def KthNode(self, pRoot, k):
        # write code here
        if not pRoot or k == 0:
            return None

        res = self.sub_tree(pRoot)
        if k > len(res):
            return None
        else:
            return res[k - 1]

    def sub_tree(self, node):
        if not node:
            return None
        res = []
        if node.left:
            res.extend(self.sub_tree(node.left))
        res.append(node)

        if node.right:
            res.extend(self.sub_tree(node.right))
        return res

if __name__ == '__main__':
    pass

File: sword_c++/二叉搜索树的第k个结点.cpp
//
// Created by 李先生 on 2018/9/4.
//
#include <iostream>
#include <vector>
using namespace std;

struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
    TreeNode(int x) :
            val(x), left(NULL), right(NULL) {
    }
};

class Solution {
public:
    vector<TreeNode *> nodevec;

    void inorder(TreeNode * pRoot)
    {
        if (pRoot == NULL)
            return;
        inorder(pRoot->left);
        nodevec.push_back(pRoot);
        inorder(pRoot->right);
    }
    TreeNode* KthNode(TreeNode* pRoot, int k)
    {
        if (k <= 0 )
            return NULL;
        inorder(pRoot);
        if (nodevec.size() > k)
            return NULL;
        return nodevec[k-1];
    }
};

int main()
{
    vector<int> res;
    cout <<res.size()<<endl;
    return 0;
}


